### DISCUSSION TITLE
Option to auto delete file when related item is deleted from collection.

### DISCUSSION THREAD

**Discussion by Posoroko** [directus/directus#17853] [Feature Requests]
### Summary

When creating a field of type "file" or "image", you have the option to auto delete the file if the image is deleted from the item, or the item is deleted from the colelction.

### Motivation

I use directus to build simple to use app for clients.  One thing I use it for is to build photo galleries.  A collection of Photo_albums in wich the user can create albums and add images in them. classic. But often, these albums are not meant to exist for a long time.  So if you delete an album from the collection, you then have to go in the file library to delete the files.  That is not very practical. And I realised that some clients won't take the time to do this. So the files are piling up and things become messy.

### Detailed Design

in the field's options:

When the image is deselected in the item:
1) delete the file
2) keep the file
3) prompt the user ("Do you want to delete the related file?")

When the item is deleted from the collection:
1) delete the file
2) keep the file
3) prompt the user ("Do you want to delete the related file?")

When the "file" or "image" field is deleted
1) delete all concerned files
2) keep all the files
3) prompt the user ("Do you want to delete the related files?")

### Must Have:
-default action for all items in collection

### Could Have:
-prompt user when an item is deleted to handle each cases differently

### Drawbacks

I suppose all features have drawbacks when it's time to implement them among existing features. It migh be super complicated without me realizing it.

### Alternatives

This can be done with flows, but it might no always be a universal solution. To "image" fields could have different names. So the flow needs to be adapted to all filed names. 

### Adoption Strategy

The default value could be "don't deleted related files".  SI it wouldn't change anything for pre-existing projects. But  I see a wide variety of cases where this would be a very useful feature.

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by w0kyj** [comment]
## Summary
Current file/image relationships only have one relational trigger option from the directus_files point of view.

This feature would allow for options to cascade delete the associated file(s) from the `collection-->field` perspective.


<img width="846" alt="File Relational Triggers Example" src="https://github.com/directus/directus/assets/40695629/ec3cf380-341e-46c9-9261-63be282a5e1d">

## Design Considerations:
- Files are not stored in the database, so API logic required to handle the storage delete
- Ensure the same file logic used from File Library when deleting files
- Set defaults as noted to non-impacting settings so current behavior is not changed

## Basic Example:
- Collection `articles` created:
    - Several fields including `title, author, body, etc`
    - File `image` field relationship created
        - set new trigger `On Delete of articles` --> `Delete the directus_files item (cascade)`
- Author has article with associated image
- The article is no longer relevant and is deleted
    - Expect the article item and related directus_files and the actual file are all deleted

**Comment by alexvdvalk** [comment]
The files could be associated to other records too, so there would need to be a way to handle this use case.

**Reply by martin-braun** [reply]
I think this really requires a routine that checks if an image has a certain upload date, so new images cannot be removed, and then checks if there are any items relating to such image to auto delete it.

**Comment by domdus** [comment]
You guys may be interested in how to achieve with a directus flow (for the moment).
https://github.com/directus/directus/discussions/6283#discussioncomment-7014994

**Comment by w0kyj** [comment]
## Request Review Notes

### Discussion
- Files create an entry in `directus_files` and store file on storage disk
- Could this apply to any type of relationship
- Things that need to be considered
    - Track where all the relationships exist --> graph style
    - Count from each of the relationships
- Workflow options:
    - When removing or deleting:
        - user prompts - to replace, nullify(only if field allows), delete
        - if delete selected - should we show other existing relationships
            - need guard rails to protect content and users from harm
- larger considerations
    - currently Directus relies on the database to handle the cascading (although performant, this is  means Directus doesn't have logic and visibility to handling these in more user friendly ways)
    - permissions to delete -- especially if related item has other associations that this user doesn't have permissions to change
    - when does delete occur on main save or immediately (top level save feels like best option)
    - ways to identify records that have no relationships

### Solution Ideas
- Need to implement some form of Directus relational handling to support this
- Want a way to identify that an item is/is not related to other collections/items
    - Layout/Item Form views
    - How to track, query, index, refresh if direct db updates done, ...

**Comment by tinhXEM** [comment]
### Simple Solution
Create duplicate files for each record in the database. This means that if the same image is used in two different records, it will be stored as two separate files (even though it’s the same image, it’s saved as two duplicates). This way, when a record is deleted, you can directly delete all related files without affecting other records.

- I know that creating duplicates of the same file isn’t the best idea, but in reality, if you’re building a system where each user’s file storage is tracked, this approach can work well (since each user’s resources are their own and they pay for what they use).
- Another scenario is from when I was developing a news website. Due to the sensitive political nature of the news, once content was approved by the editor-in-chief, it couldn’t be modified anymore. So even if a journalist uploaded the same image again, it created a new copy to avoid one person accidentally overwriting another’s image, which could distort the content.

### The Ideal Solution

- Keep a record of which resources (files) are being used by each record (including the collection name and the record ID). Whenever a record is added, edited, or deleted, the system should check all resources used in that record to ensure everything is updated correctly (compare before submitting the changes and after submission to keep track).
- When deleting, a confirmation prompt will appear with a link to the related items so the user can view them and decide whether to delete the file or not.

**Comment by Posoroko** [comment]
I made that feature request 2 years ago and I just realized that you addressed it in a request review stream video on sept 12th 2024.
Here is a bit more context and about the motivations for this feature.

The reason why I made this request is based on how I use files in directus,, witch is not how the "default" way proposed by directus.

In directus, files are assets that the project owns and that users can use. Pretty much like files on our computers.

But the way I was/am using them is this:
User post a classified ad to sell a car and adds a photo of the car. When the car is sold, the item is deleted and the image is no longer usefull.

so 2 different senarios :

    "items depend on file"
    item.image is m2o field
    files is part of a collection available to the users.

    "files depend on item"
    item.images is o2m
    files only exist in relation to the item.

these are really different, but I feel option 2 is a valid way to think about it in many user content apps.

Thanks for addressing my request ! I still use this concept a lot, so if I can help by sharing m experience, I'd be glad to do so !

**Reply by ChuckMoe** [reply]
I have the exact same use case. Someone posts an ad with images and as soon as that user generated data is gone, I want the image data to be deleted as well.

**Comment by Heliumrich** [comment]
How to do this ?
Is there some extension for it ? can't find any

I could AI slop my way out with some node.js or PostgreSQL queries and try to automate it with systemd or something but...

**Reply by kheiner** [reply]
If I were tasked with setting this up, I'd look into using [Flows](https://directus.com/docs/guides/flows).

> Flows enable custom, event-driven data processing and task automation within Directus. Each flow is composed of one trigger, a series of operations, and a data chain that is passed between each step.

Let me know if that answers the question or gives you a good starting point. Happy to continue to help if you need it.

**Reply by domdus** [reply]
See: https://github.com/directus/directus/discussions/6283#discussioncomment-7014994

**Reply by Heliumrich** [reply]
This seems to only work for new images I deselect, but it won't work for images I already deselected in the past, right ?

Is there any way to just do once, manually, for images not used right now ? Like a periodic clean or something

**Reply by kheiner** [reply]
> This seems to only work for new images I deselect, but it won't work for images I already deselected in the past, right ?

Correct.

> Is there any way to just do once, manually, for images not used right now ? Like a periodic clean or something

Anything's possible. Are you asking for help with how to do it or just if it's something that's feasible? If you set up the flow as provided in the example above you'd get a go-forward clean and all you'd need is a one time cleanup (no need for a periodic clean).



See the docs for [API](https://directus.com/docs/api/files#delete-multiple-files), [Queries](https://directus.com/docs/guides/connect/query-parameters) and [Filters](https://directus.com/docs/guides/connect/filter-rules#available-operators). Hopefully that helps.

**Reply by Heliumrich** [reply]
Well it's cool but no way I am checking 500 items, deleting one by one the 300-800 unused images (I changed most of them once or twice)

Or download all 500 used images, delete everything, and reupload each on the correct item...

Guess I'll try to AI slop my way out of this. Surely there's a way to get the filename used from the DB, and then delete everything except these

**Reply by domdus** [reply]
Due to the talks above I implemented this feature now in `Storage Manager` extension.
Marketplace or https://github.com/domdus/directus-extension-storage-manager -> #unreferenced-files.

Actually I was a bit scared to post it here 😅, but I tested intensively to make sure it works reliable also with big amounts of files. I am using on a production env with more than a million of files.

The bundle extension comes with a lot more related to file cleanup such as  settings for deselect files behavior in `native file interfaces` (delete or keep), extended  `file interfaces with storage`, `recycle bin`, flow `operations` etc., see: README.md.

A possible work flow:

1. Scan for unreferenced files that are used nowhere, not in file fields, not in relations, not even as loose uids in text fields (JSON/WYSIWYG)
2. If you hesitate to delete the identified files right after scan -> enable Recycle Bin feature and move the files into that folder
3. Files in Recycle Bin are isolated -> thumbnails/transformations are deleted and the file is not served via /assets endpoint anymore. 
4. (!) In case of false positives (should not happen) this will be the point were your product (website/app/api service) starts to show broken images, missing resources, or your editors notice missing images even in the Directus Studio. If it turns out that files are actually still needed/referenced -> move them out anywhere else and they will become available again. Nothing is lost. (Rollback)
5. With the default retention -> recycled files are going to be deleted after 30 days on `manual purge of expired`
6. Enable Scheduled Purge in order to daily purge the Recycle Bin for the configured retention

Please report bugs!

---------------------------------------------------------------------

<img width="400" height="auto" alt="image" src="https://github.com/user-attachments/assets/07e7f728-5adf-4fcd-9f40-87e997f6f1e6" />
<img width="400" height="auto" alt="image" src="https://github.com/user-attachments/assets/f27cf29b-a747-42dc-8cc8-898b3d5b8cf4" />
